core-bundle/src/Security/Voter/DataContainer/PagePermissionVoter.php

```diff
@@ -0,0 +1,233 @@
+    ) {
+    }
+
+    public function supportsAttribute(string $attribute): bool
+    {
+        return \in_array($attribute, [ContaoCorePermissions::DC_PREFIX.'tl_page', ContaoCorePermissions::DC_PREFIX.'tl_article'], true);
+    }
+
+    public function supportsType(string $subjectType): bool
+    {
+        return \in_array($subjectType, [CreateAction::class, ReadAction::class, UpdateAction::class, DeleteAction::class], true);
+    }
+
+    public function reset(): void
+    {
+        $this->pagemountsCache = [];
+    }
+
+    public function vote(TokenInterface $token, $subject, array $attributes): int
+    {
+        if (!array_filter($attributes, $this->supportsAttribute(...))) {
+            return self::ACCESS_ABSTAIN;
+        }
+
+        if ($this->accessDecisionManager->decide($token, ['ROLE_ADMIN'])) {
+            return self::ACCESS_ABSTAIN;
+        }
+
+        $isGranted = match (true) {
+            $subject instanceof CreateAction => $this->canCreate($subject, $token),
+            $subject instanceof ReadAction => $this->canRead($subject, $token),
+            $subject instanceof UpdateAction => $this->canUpdate($subject, $token),
+            $subject instanceof DeleteAction => $this->canDelete($subject, $token),
+            default => null,
+        };
+
+        if (false === $isGranted) {
+            return self::ACCESS_DENIED;
+        }
+
+        return self::ACCESS_ABSTAIN;
+    }
+
+    private function canCreate(CreateAction $action, TokenInterface $token): bool
+    {
+        // The copy operation is allowed if either hierarchy or edit is allowed.
+        if (null !== $action->getNew() && null === ($action->getNew()['sorting'] ?? null)) {
+            $pageId = match ($action->getDataSource()) {
+                'tl_page' => (int) $action->getNewId(),
+                'tl_article' => (int) $action->getNewPid(),
+                default => throw new \UnexpectedValueException(sprintf('Unsupported data source "%s"', $action->getDataSource())),
+            };
+
+            return ($this->canEdit($action, $token, $pageId) || $this->canChangeHierarchy($action, $token, $pageId))
+                && $this->canAccessPage($token, $pageId)
+                && $this->canCreate(new CreateAction($action->getDataSource()), $token);
+        }
+
+        // Check access to any page on `new` operation
+        if (null === $action->getNewPid()) {
+            $pageIds = $this->getPagemounts($token);
+        } else {
+            $pageIds = [(int) $action->getNewPid()];
+        }
+
+        // To actually create a record, both hierarchy and edit permissions must be available.
+        foreach ($pageIds as $pageId) {
+            if (
+                $this->canEdit($action, $token, $pageId)
+                && $this->canChangeHierarchy($action, $token, $pageId)
+                && $this->canAccessPage($token, $pageId)
+            ) {
+                return true;
```

Comment:
This breaks the loop at the first page ID that satisfies the requirements of the if condition, which makes it _allow if any page is allowed_. But should it not be _allow if all pages are allowed_ instead?

Reply:
If `$action->getNewPid()` is set, the array has only one item so the logic is correct for that case.
If it is empty (meaning a create action without a pid), `$pageIds` gets set to the pagemounts. In that case if creating is allowed in any of the mounted pages than this method should return true as “create” (without the specified pid) is allowed then.